compiler/rustc_lint/src/builtin.rs

```diff
@@ -3098,3 +3094,54 @@ impl<'tcx> LateLintPass<'tcx> for DerefNullPtr {
         }
     }
 }
+
+declare_lint! {
+    /// The `const_transmute_ptr_to_number` lint detects transmutes from a
+    /// pointer to a number in a `const` context.
+    ///
+    ///
+    /// ### Example
+    ///
+    /// ```rust,no_run
+    /// # #![allow(dead_code)]
+    /// pub const fn x() -> usize {
+    ///     unsafe { std::mem::transmute(&0) }
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Transmuting a pointer to a number may be [undefined behavior]. See
+    /// [this issue] for details.
+    ///
+    /// [undefined behavior]: https://github.com/rust-lang/unsafe-code-guidelines/issues/286[
+    /// [this issue]: https://github.com/rust-lang/unsafe-code-guidelines/issues/286
+    pub CONST_TRANSMUTE_PTR_TO_NUMBER,
+    Deny,
+    "detects when a pointer is transmuted into a number in a const context"
+}
+
+declare_lint_pass!(ConstTransmutePtrToNumber => [CONST_TRANSMUTE_PTR_TO_NUMBER]);
+
+impl<'tcx> LateLintPass<'tcx> for ConstTransmutePtrToNumber {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &hir::Expr<'_>) {
+        if !cx.tcx.hir().is_inside_const_context(expr.hir_id) {
+            return;
+        }
+
+        if let Some((
+            &ty::RawPtr(..) | &ty::Ref(..),
+            &ty::Int(..) | &ty::Uint(..) | &ty::Float(..),
+        )) = get_transmute_from_to(cx, expr).map(|(ty1, ty2)| (ty1.kind(), ty2.kind()))
+        {
+            cx.struct_span_lint(CONST_TRANSMUTE_PTR_TO_NUMBER, expr.span, |lint| {
+                lint.build("cast of pointer to number")
+                    .note("this may be undefined behavior")
+                    .help("see <https://github.com/rust-lang/unsafe-code-guidelines/issues/286> for details")
```

Comment:
Same comment about the link. If we link somewhere it should be specific about these transmutes in `const` context.